clusterloader2/pkg/provider/provider.go

```diff
@@ -0,0 +1,108 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"fmt"
+	"strings"
+)
+
+type InitOptions struct {
+	KubemarkRootKubeConfigPath string
+	ProviderName               string
+	ProviderConfigs            []string
+}
+
+type Features struct {
+	// SupportWindowsNodeScraping determines wheter scraping windows node in supported.
+	SupportWindowsNodeScraping bool
+	// SupportProbe determines whether probe is supported.
+	SupportProbe bool
+	// SupportImagePreload determines whether image preloading is supported.
+	SupportImagePreload bool
+	// SupportEnablePrometheusServer determines whether enabling prometheus server is possible.
+	SupportEnablePrometheusServer bool
+	// SupportSSHToMaster determines whether SSH access to master machines is possible.
+	// If false (impossible for many  providers), ClusterLoader will skip operations requiring it.
+	SupportSSHToMaster bool
+	// SupportAccessAPIServerPprofEndpoint determines whether accessing api server pprof endpoint is possible.
+	SupportAccessAPIServerPprofEndpoint bool
+	// SupportSnapshotPrometheusDisk determines whether snapshot prometheus disk is supported.
+	SupportSnapshotPrometheusDisk bool
+	// SupportNodeKiller determines whether node killer is supported.
+	SupportNodeKiller bool
+	// SupportGrabMetricsFromKubelets determines whether getting metrics from kubelet is supported.
+	SupportGrabMetricsFromKubelets bool
+
+	// ShouldPrometheusScrapeApiserverOnly determines if we should set PROMETHEUS_SCRAPE_APISERVER_ONLY by default.
+	ShouldPrometheusScrapeApiserverOnly bool
+}
+
+type Provider interface {
+	// Name returns name of this provider. It should used only in logs.
+	Name() string
+	// Features returns the feature supported by this provider.
+	Features() *Features
+
+	GetPortForComponent(componentName string) (int, error)
+	GetProtocolForComponent(componentName string) string
+	GetRootFrameworkKubeConfigOverride() string
```

Comment:
What is this? Why is this needed in this interface?

Reply:
I see that now - it's here: https://github.com/kubernetes/perf-tests/pull/1344/files#diff-7da509c349937d31912207644b69bc81R34

Given it's actually one of fields in config, I would like to avoid these fancy fields (this may be growing indefinitely in the future). If we really need that (and it seems we do), let's do:
```
type Config map[string]string

// in this interface expose the config
  GetConfig() Config

// And then you can expose some methods on the config as helpers, e.g.
func (c *Config) RootFrameworkKubeConfigOverride() string {
  return c[RootKubeConfigKey]
}
```